### ISSUE TITLE
Bug in ConstraintMatrix::add_entries()

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#7101]
As reported on the mailing list by Dustin Kumor:
.....................................
Dear all,

in my opinion there is a small bug in the implementation of the ConstraintMatrix member function add_entries in file constraint_matrix.cc (release Version 9.0, lines 224-261). The current piece of code is:
```
  224 void
  225 ConstraintMatrix::add_entries
  226 (const size_type                                  line,
  227  const std::vector<std::pair<size_type,double> > &col_val_pairs)
  228 {
  229   Assert (sorted==false, ExcMatrixIsClosed());
  230   Assert (is_constrained(line), ExcLineInexistant(line));
  231
  232   ConstraintLine *line_ptr = &lines[lines_cache[calculate_line_index(line)]];
  233   Assert (line_ptr->index == line, ExcInternalError());
  234
  235   // if in debug mode, check whether an entry for this column already
  236   // exists and if its the same as the one entered at present
  237   //
  238   // in any case: skip this entry if an entry for this column already
  239   // exists, since we don't want to enter it twice
  240   for (std::vector<std::pair<size_type,double> >::const_iterator
  241        col_val_pair = col_val_pairs.begin();
  242        col_val_pair!=col_val_pairs.end(); ++col_val_pair)
  243     {
  244       Assert (line != col_val_pair->first,
  245               ExcMessage ("Can't constrain a degree of freedom to itself"));
  246
  247       for (ConstraintLine::Entries::const_iterator
  248            p=line_ptr->entries.begin();
  249            p != line_ptr->entries.end(); ++p)
  250         if (p->first == col_val_pair->first)
  251           {
  252             // entry exists, break innermost loop
  253             Assert (p->second == col_val_pair->second,
  254                     ExcEntryAlreadyExists(line, col_val_pair->first,
  255                                           p->second, col_val_pair->second));
  256             break;
  257           }
  258
  259       line_ptr->entries.push_back (*col_val_pair);
  260     }
  261 }
```
Although, it is stated in the comments (lines 235-239) there is no skip of the entry col_val_pair in the case this entry already exist. The break command terminates the innermost loop but the push_back operation is performed nevertheless, i.e. for every element of col_val_pairs. This leads to duplicates in the constraint matrix.
I solved the problem by just addind the three lines 246a, 255a, 258a.
```
 224 void
 225 ConstraintMatrix::add_entries
 226 (const size_type                                  line,
 227  const std::vector<std::pair<size_type,double> > &col_val_pairs)
 228 {
 229   Assert (sorted==false, ExcMatrixIsClosed());
 230   Assert (is_constrained(line), ExcLineInexistant(line));
 231
 232   ConstraintLine *line_ptr = &lines[lines_cache[calculate_line_index(line)]];
 233   Assert (line_ptr->index == line, ExcInternalError());
 234
 235   // if in debug mode, check whether an entry for this column already
 236   // exists and if its the same as the one entered at present
 237   //
 238   // in any case: skip this entry if an entry for this column already
 239   // exists, since we don't want to enter it twice
 240   for (std::vector<std::pair<size_type,double> >::const_iterator
 241        col_val_pair = col_val_pairs.begin();
 242        col_val_pair!=col_val_pairs.end(); ++col_val_pair)
 243     {
 244       Assert (line != col_val_pair->first,
 245               ExcMessage ("Can't constrain a degree of freedom to itself"));
 246
 246a      bool entry_exists = false;
 247       for (ConstraintLine::Entries::const_iterator
 248            p=line_ptr->entries.begin();
 249            p != line_ptr->entries.end(); ++p)
 250         if (p->first == col_val_pair->first)
 251           {
 252             // entry exists, break innermost loop
 253             Assert (p->second == col_val_pair->second,
 254                     ExcEntryAlreadyExists(line, col_val_pair->first,
 255                                           p->second, col_val_pair->second));
 255a            entry_exists = true;
 256             break;
 257           }
 258
 258a       if (entry_exists == fase)
 259           line_ptr->entries.push_back (*col_val_pair);
 260     }
 261 }
```
May be there is a more elegant way to do this.

Best regards
Dustin

**Comment by dkumorAtGitHub** [comment]
Hallo everyone,

within the conversation on the mailing list Wolfgang Bangerth also proposed to write a test. At the moment I'm not really sure how this test should look like. The only thing I could imagine would be to compare a constraint matrix filled with the add_entry () member function to one filled with the same entries by using the add_entries () member function. However, I do not know if this is a reasonable test.

**Comment by drwells** [comment]
That sounds reasonable: all we need to do is make sure that, with #7114, no duplicate entries are added.

**Comment by masterleinad** [comment]
Fixed in #7114.
